libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/OAuthWebviewHelper.kt

```diff
@@ -661,7 +661,7 @@ open class OAuthWebviewHelper : KeyChainAliasCallback {
                     )
 
                     else -> when {
-                        instance.useWebServerAuthentication -> onWebServerFlowComplete(params["code"])
+                        codeVerifier?.isNotEmpty() == true -> onWebServerFlowComplete(params["code"])
```

Comment:
This is a bug and needs to be reverted as it will break User Agent flow login.  `codeVerifier` is always **created** no matter which flow we are using.

Reply:
You are right, I did not try user agent flow when doing the spike.
But we can't just revert.
With QR login, we will decide what flow to use based on what came through the JSON encoded in the QR URL.
If the JSON contains a code verifier, we use it and we should to a web server flow. And we should set a new flag saying that web server flow should be used.
That flag should be set to instance.useWebServerAuthentication by default.
We should use the new flag here.

Reply:
Thanks, @brandonpage!  I'll take a look.

Reply:
Okay.  So if I understand correctly the QR URL can/should override the app's setting for `useWebServerAuthentication`?  Sounds like something that should be documented. 

Also is User Agent Flow supported?  The code above appears to require PKCE.

Reply:
We support both, if the QR encodes a code verifier we expect to the bridge URL to do a web server flow and the redirect to get us a code. If the QR comes with an empty/null code verifier, then the code verifier ends up not being defined and the code above will flow the user agent branch. The bridge URL should do a user agent flow and the redirect get us the refresh tokens.

During the spike, I played with both, but the web server flow case meant a longer URL encoded in the QR which made it harder to read with the reader. I could not get it to work completely. See https://salesforce.quip.com/JXmvAwirhR3V#temp:C:IIO55768e89acdc49948a3be45bf

Reply:
@JohnsonEricAtSalesforce Before closing the story, we need to make sure user agent and web server flow work with and without QR. During the spike, I did not test them all or get them all working.

Reply:
@wmathurin & @brandonpage - I broke out [W-16336715: [MSDK Android] Correct QR Code Log In When Using Web Server](https://gus.lightning.force.com/a07EE00001xT1B3YAK) to track this in detail.  I'll commit that here before we merge for 12.1 or whichever release QR code login lands in soon.

I also broke out [W-16336722](https://gus.lightning.force.com/a07EE00001xTbt7YAC) to track the other bug in Quip regarding the main activity not being displayed.